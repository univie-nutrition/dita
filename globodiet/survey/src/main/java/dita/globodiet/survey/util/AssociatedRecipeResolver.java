/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package dita.globodiet.survey.util;

import java.math.BigDecimal;
import java.util.Optional;

import org.apache.causeway.commons.collections.Can;
import org.apache.causeway.commons.internal.exceptions._Exceptions;

import lombok.NonNull;

import dita.commons.food.consumption.FoodConsumption.ConsumptionUnit;
import dita.commons.format.FormatUtils;
import dita.commons.sid.SemanticIdentifier;
import dita.commons.sid.SemanticIdentifier.SystemId;
import dita.commons.sid.SemanticIdentifierSet;
import dita.foodon.fdm.FoodDescriptionModel;
import dita.foodon.fdm.FoodDescriptionModel.Recipe;
import dita.recall24.dto.RecallNode24;
import dita.recall24.dto.RecallNode24.Annotation;
import dita.recall24.dto.RecallNode24.Transfomer;
import dita.recall24.dto.Record24;
import dita.recall24.dto.Record24.Composite;
import dita.recall24.dto.Record24.Food;

/**
 * In 2024, the Austrian GloboDiet food description model introduced the concept of 'associated' food,
 * that can be quickly entered during interviews,
 * but requires a post-processing step later (before interview reporting),
 * with which those reported food nodes are to be replaced by an associated recipe.
 * <p>
 * The association link between food and recipe is made via references in their name:
 * <ul>
 * <li>food reference their associated recipe via suffix {assocRecp=‹recipe-code›}</li>
 * <li>recipes reference their associated food via suffix {assocFood=‹food-code›}</li>
 * </ul>
 */
public record AssociatedRecipeResolver(
        @NonNull FoodDescriptionModel foodDescriptionModel
        ) implements Transfomer {

    @SuppressWarnings("unchecked")
    @Override
    public <T extends RecallNode24> T transform(final T node) {

        return switch(node) {
            // we replace the original food by its associated recipe
            case Food origFood -> {

                var fdmAdapter = new FDMAdapter(origFood.sid().systemId(), foodDescriptionModel);

                var associatedRecipe = fdmAdapter
                        .associatedRecipe(NameWithCode.parseAssocRecipe(origFood.name()))
                        .orElse(null);
                if(associatedRecipe==null) yield (T)origFood;

                var recordBuilder = new Composite.Builder();

                // replace the (proxy-) food node by its associated composite node ..

                recordBuilder.type(Record24.Type.COMPOSITE);
                recordBuilder.name(NameWithCode.parseAssocFood(associatedRecipe.name()).nameWithResolvedSuffix());
                recordBuilder.sid(associatedRecipe.sid());
                // keep the original food as comment
                recordBuilder.subRecords().add(origFoodAsComment(origFood));

                // store GloboDiet food description group data as annotation
                recordBuilder.annotations().clear();
                recordBuilder.annotations().add(new Annotation("group", recipeGroupSid(associatedRecipe)));

                //TODO[dita-globodiet-survey-24] what facets to put here?
                recordBuilder.facetSids(SemanticIdentifierSet.wip());

                var origFoodConsumedOverRecipeMass = new BigDecimal(
                        origFood.amountConsumed().doubleValue() //TODO[dita-globodiet-survey] might not always be in GRAM
                        / foodDescriptionModel.sumAmountGramsForRecipe(associatedRecipe).doubleValue());

                // to the composite record we are building here,
                // we add each recipe ingredient as sub-record
                foodDescriptionModel.streamIngredients(associatedRecipe)
                    .map(ingr->{
                        var food = foodDescriptionModel.lookupFoodBySid(ingr.foodSid()).orElseThrow();
                        var foodBuilder = new Food.Builder()
                            .name(food.name())
                            .sid(ingr.foodSid())
                            .facetSids(SemanticIdentifierSet.wip()) //TODO[dita-globodiet-survey] missing ingredient facets
                            .amountConsumed(ingr.amountGrams().multiply(origFoodConsumedOverRecipeMass))
                            .consumptionUnit(ConsumptionUnit.GRAM); // for recipes this is always in GRAM
                        foodBuilder.annotations().add(new Annotation("group", foodGroupSid(food)));
                        return foodBuilder.build();
                    })
                    .forEach(recordBuilder.subRecords()::add);

                yield (T)recordBuilder.build();

            }
            default -> node;
        };
    }

    // -- HELPER

    private SemanticIdentifier foodGroupSid(final FoodDescriptionModel.Food food) {
        return food.groupSid()
                .mapObjectId(o->o.mapContext(_->SidUtils.GdContext.FOOD_GROUP.id()));
    }

    private SemanticIdentifier recipeGroupSid(final FoodDescriptionModel.Recipe recipe) {
        return recipe.groupSid()
                .mapObjectId(o->o.mapContext(_->SidUtils.GdContext.RECIPE_GROUP.id()));
    }

    private Record24.Comment origFoodAsComment(final Food origFood) {
        var name = "%s, amount=%s, raw/cooked=%s".formatted(
                origFood.name().replace(" {", ", ").replace("}", ""),
                origFood.consumptionUnit().format(origFood.amountConsumed()),
                origFood.rawPerCookedRatio());

        return Record24.comment(name, origFood.sid(), origFood.facetSids(),
                Can.ofCollection(origFood.annotations().values()));
    }

    private record FDMAdapter(
            @NonNull SystemId systemId,
            @NonNull FoodDescriptionModel foodDescriptionModel) {

        /**
         * Optionally returns the associated {@link Recipe},
         * based on whether a suffix {assocRecp=‹recipe-code›} is present.
         * <p>
         * Fails if the recipe-code is present but cannot be resolved.
         */
        Optional<Recipe> associatedRecipe(@NonNull final NameWithCode nameWithCode) {
            var associatedRecipeSid = nameWithCode.associatedRecipeSid(systemId).orElse(null);
            if(associatedRecipeSid == null) return Optional.empty();
            var associatedRecipe = foodDescriptionModel.lookupRecipeBySid(associatedRecipeSid);
            if(associatedRecipe.isEmpty()) {
                throw _Exceptions.illegalArgument("failed to resolve recipe for %s using sid %s",
                        nameWithCode.name(),
                        associatedRecipeSid);
            }
            return associatedRecipe;
        }

    }

    private record NameWithCode(String name, String code) {
        static NameWithCode parseAssocFood(final String nameAndCode) {
            return parse(nameAndCode, "{assocFood=");
        }
        static NameWithCode parseAssocRecipe(final String nameAndCode) {
            return parse(nameAndCode, "{assocRecp=");
        }
        static NameWithCode parse(final String nameAndCode, final String magic) {
            int c1 =nameAndCode.indexOf(magic);
            if(c1==-1) return new NameWithCode(nameAndCode, null);
            int c2 = nameAndCode.indexOf("}", c1);
            return new NameWithCode(
                    nameAndCode.substring(0, c1).trim(),
                    FormatUtils.fillWithLeadingZeros(5, nameAndCode.substring(c1 + magic.length(), c2)));
        }
        String nameWithResolvedSuffix() {
            return name() + " {resolved}";
        }
        private Optional<SemanticIdentifier> associatedRecipeSid(final SystemId systemId) {
            return Optional.ofNullable(code())
                    .map(code->SidUtils.GdContext.RECIPE.sid(systemId, code));
        }
    }

}
