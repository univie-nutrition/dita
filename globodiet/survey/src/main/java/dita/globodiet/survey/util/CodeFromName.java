/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package dita.globodiet.survey.util;

import java.util.Optional;

import org.jspecify.annotations.Nullable;

import org.springframework.util.StringUtils;

import org.apache.causeway.commons.internal.base._Strings;
import org.apache.causeway.commons.io.TextUtils;

import dita.commons.sid.SemanticIdentifier;
import dita.commons.sid.SemanticIdentifier.ObjectId;
import dita.commons.sid.SemanticIdentifier.SystemId;
import dita.commons.util.FormatUtils;

record CodeFromName(@Nullable String code) {

    static CodeFromName parseAssocFood(final String nameAndCode) {
        return parse("assocFood", nameAndCode);
    }
    static CodeFromName parseAssocRecipe(final String nameAndCode) {
        return parse("assocRecp", nameAndCode);
    }
    private static CodeFromName parse(final String attributeName, final String nameAndCode) {
        int c1 = nameAndCode.indexOf("{");
        return (c1==-1)
            ? new CodeFromName(null)
            : new CodeFromName(
                lookupAttribute(attributeName, nameAndCode)
                    .map(code->FormatUtils.fillWithLeadingZeros(5, code))
                    .orElse(null));
    }
    Optional<SemanticIdentifier> associatedRecipeSid(final SystemId systemId) {
        return Optional.ofNullable(code())
                .map(code->ObjectId.Context.RECIPE.sid(systemId, code));
    }

    /// looks for `value` in `{.., key=value,..}`
    private static Optional<String> lookupAttribute(final String attributeKey, final String curlybracedAttributeList) {
        final String commaSeparatedKeyValuePairs = TextUtils.cutter(curlybracedAttributeList)
            .keepAfter("{")
            .keepBeforeLast("}")
            .getValue();
        final String pattern = attributeKey + "=";

        return _Strings.splitThenStream(commaSeparatedKeyValuePairs, ",")
            .map(String::trim)
            .filter(kv->kv.startsWith(pattern))
            .map(kv->kv.substring(pattern.length()))
            .filter(StringUtils::hasLength)
            .findFirst();
    }

}

