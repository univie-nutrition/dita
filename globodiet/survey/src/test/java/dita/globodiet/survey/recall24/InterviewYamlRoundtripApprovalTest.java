/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package dita.globodiet.survey.recall24;

import org.approvaltests.Approvals;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.UseReporter;
import org.junit.jupiter.api.Test;

import dita.commons.sid.SemanticIdentifier.SystemId;
import dita.recall24.dto.InterviewSet24;
import dita.recall24.dto.util.InterviewSetParser;
import dita.testing.ApprovalTestOptions;

class InterviewYamlRoundtripApprovalTest {

    private SystemId systemId = new SystemId("at.gd", "2.0");

    @Test
    @UseReporter(DiffReporter.class)
    void parsingSample() {
        var interviewSet = InterviewSamples.SAMPLES.asInterviewSet(systemId, null);
        Approvals.verify(rountrip(interviewSet).toYaml(), ApprovalTestOptions.yamlOptions());
    }

    @Test
    @UseReporter(DiffReporter.class)
    void parsingRefComposites() {
        var interviewSet = InterviewSamples.COMPOSITES.asInterviewSet(systemId, null);
        Approvals.verify(rountrip(interviewSet).toYaml(), ApprovalTestOptions.yamlOptions());
    }

    @Test
    @UseReporter(DiffReporter.class)
    void parsingRefFatSouceSweetener() {
        var interviewSet = InterviewSamples.FAT_SOUCE_SWEETENERS.asInterviewSet(systemId, null);
        Approvals.verify(rountrip(interviewSet).toYaml(), ApprovalTestOptions.yamlOptions());
    }
    
    // -- HELPER
    
    static InterviewSet24 rountrip(InterviewSet24 in) {
        return InterviewSetParser.parseYaml(in.toYaml());
    }

}
