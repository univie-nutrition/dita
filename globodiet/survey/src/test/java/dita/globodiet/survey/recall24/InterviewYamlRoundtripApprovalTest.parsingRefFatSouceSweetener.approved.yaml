respondents:
- alias: "1234"
  dateOfBirth: "2000-01-01"
  sex: "FEMALE"
  interviews:
  - interviewDate: "2024-05-29"
    consumptionDate: "2024-05-28"
    interviewOrdinalRef:
      value: 1
    respondentSupplementaryData:
      specialDietId: "00"
      specialDayId: "00"
      heightCM: 160
      weightKG: 60
    meals:
    - hourOfDay: "09:00:00"
      foodConsumptionOccasionId: "02"
      foodConsumptionPlaceId: "01"
      memorizedFood:
      - name: "Kaffee, mit Koffein und Zucker"
        topLevelRecords:
        - type: "FOOD"
          name: "Kaffee, mit Koffein {assocRecp=492}"
          sid: "at.gd/2.0:food/01339"
          facetSids: "at.gd/2.0:fd/0201, at.gd/2.0:fd/0699, at.gd/2.0:fd/0701, at.gd/2.0:fd/1200"
          amountConsumed: 150
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/130301"
        - type: "FOOD"
          name: "Zucker, Kristall"
          sid: "at.gd/2.0:food/00992"
          facetSids: "at.gd/2.0:fd/1200"
          amountConsumed: 7.5
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/110101"
    - hourOfDay: "12:00:00"
      foodConsumptionOccasionId: "05"
      foodConsumptionPlaceId: "01"
      memorizedFood:
      - name: "Salat, gemischt mit Essig/Öl-Dressing"
        topLevelRecords:
        - type: "FOOD"
          name: "Salat, Eichblatt"
          sid: "at.gd/2.0:food/00036"
          facetSids: "at.gd/2.0:fd/0204, at.gd/2.0:fd/0499, at.gd/2.0:fd/0910, at.gd/2.0:fd/2807"
          amountConsumed: 20
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/0201"
        - type: "FOOD"
          name: "Salat, Kopf n.s."
          sid: "at.gd/2.0:food/00040"
          facetSids: "at.gd/2.0:fd/0204, at.gd/2.0:fd/0499, at.gd/2.0:fd/0910, at.gd/2.0:fd/2807"
          amountConsumed: 20
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/0201"
        - type: "FOOD"
          name: "Salat, Romana"
          sid: "at.gd/2.0:food/00031"
          facetSids: "at.gd/2.0:fd/0204, at.gd/2.0:fd/0499, at.gd/2.0:fd/0910, at.gd/2.0:fd/2807"
          amountConsumed: 20
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/0201"
        - type: "FOOD"
          name: "Salatdressing, Essig/Öl {assocRecp=384}"
          sid: "at.gd/2.0:food/01527"
          facetSids: "at.gd/2.0:fd/0499, at.gd/2.0:fd/0699, at.gd/2.0:fd/1009, at.gd/2.0:fd/1200"
          amountConsumed: 12
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/150102"
      - name: "Salat, gemischt mit Joghurtdressing (Rezept bekannt)"
        topLevelRecords:
        - type: "FOOD"
          name: "Salat, Eichblatt"
          sid: "at.gd/2.0:food/00036"
          facetSids: "at.gd/2.0:fd/0204, at.gd/2.0:fd/0499, at.gd/2.0:fd/0910, at.gd/2.0:fd/2807"
          amountConsumed: 20
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/0201"
        - type: "FOOD"
          name: "Salat, Kopf n.s."
          sid: "at.gd/2.0:food/00040"
          facetSids: "at.gd/2.0:fd/0204, at.gd/2.0:fd/0499, at.gd/2.0:fd/0910, at.gd/2.0:fd/2807"
          amountConsumed: 20
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/0201"
        - type: "FOOD"
          name: "Salat, Romana"
          sid: "at.gd/2.0:food/00031"
          facetSids: "at.gd/2.0:fd/0204, at.gd/2.0:fd/0499, at.gd/2.0:fd/0910, at.gd/2.0:fd/2807"
          amountConsumed: 20
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/0201"
        - type: "FOOD"
          name: "Salatdressing, Jogurt {assocRecp=334}"
          sid: "at.gd/2.0:food/01525"
          facetSids: "at.gd/2.0:fd/0201, at.gd/2.0:fd/0499, at.gd/2.0:fd/0699, at.gd/2.0:fd/1001,\
            \ at.gd/2.0:fd/1501, at.gd/2.0:fd/1601"
          amountConsumed: 12
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking:
            type: "TYPE_OF_FAT_USED"
            name: "Öl, Raps"
            sid: "at.gd/2.0:food/00937"
            facetSids: "at.gd/2.0:fd/1200"
          typeOfMilkOrLiquidUsedDuringCooking:
            type: "TYPE_OF_MILK_OR_LIQUID_USED"
            name: "Jogurt"
            sid: "at.gd/2.0:food/00236"
            facetSids: "at.gd/2.0:fd/0102, at.gd/2.0:fd/0699, at.gd/2.0:fd/0701, at.gd/2.0:fd/0801,\
              \ at.gd/2.0:fd/1200"
          annotations:
            group: "at.gd/2.0:fg/150102"
      - name: "Salat, gemisch mit Joghurtdressing (zubereitet in Rest.)"
        topLevelRecords:
        - type: "FOOD"
          name: "Salat, Eichblatt"
          sid: "at.gd/2.0:food/00036"
          facetSids: "at.gd/2.0:fd/0204, at.gd/2.0:fd/0499, at.gd/2.0:fd/0910, at.gd/2.0:fd/2807"
          amountConsumed: 20
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/0201"
        - type: "FOOD"
          name: "Salat, Kopf n.s."
          sid: "at.gd/2.0:food/00040"
          facetSids: "at.gd/2.0:fd/0204, at.gd/2.0:fd/0499, at.gd/2.0:fd/0910, at.gd/2.0:fd/2807"
          amountConsumed: 20
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/0201"
        - type: "FOOD"
          name: "Salat, Romana"
          sid: "at.gd/2.0:food/00031"
          facetSids: "at.gd/2.0:fd/0204, at.gd/2.0:fd/0499, at.gd/2.0:fd/0910, at.gd/2.0:fd/2807"
          amountConsumed: 20
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/0201"
        - type: "FOOD"
          name: "Salatdressing, Jogurt {assocRecp=334}"
          sid: "at.gd/2.0:food/01525"
          facetSids: "at.gd/2.0:fd/0201, at.gd/2.0:fd/0400, at.gd/2.0:fd/0699, at.gd/2.0:fd/1009,\
            \ at.gd/2.0:fd/1200"
          amountConsumed: 12
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/150102"
      - name: "Pommes frites mit Ketchup"
        topLevelRecords:
        - type: "FOOD"
          name: "Pommes frites"
          sid: "at.gd/2.0:food/00010"
          facetSids: "at.gd/2.0:fd/0204, at.gd/2.0:fd/0310, at.gd/2.0:fd/0499, at.gd/2.0:fd/0801,\
            \ at.gd/2.0:fd/1200"
          amountConsumed: 142
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/0101"
        - type: "FRYING_FAT"
          name: "Öl, Raps"
          sid: "at.gd/2.0:food/00937"
          facetSids: "at.gd/2.0:fd/1200"
          amountConsumed: 25.560001373291016
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          annotations: {}
        - type: "FOOD"
          name: "Ketchup"
          sid: "at.gd/2.0:food/01515"
          facetSids: "at.gd/2.0:fd/0699, at.gd/2.0:fd/1200"
          amountConsumed: 28.399999618530273
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/150101"
        - type: "FOOD"
          name: "Pommes frites"
          sid: "at.gd/2.0:food/00010"
          facetSids: "at.gd/2.0:fd/0204, at.gd/2.0:fd/0309, at.gd/2.0:fd/0402, at.gd/2.0:fd/0801,\
            \ at.gd/2.0:fd/1200"
          amountConsumed: 142
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/0101"
        - type: "FRYING_FAT"
          name: "Öl, Raps"
          sid: "at.gd/2.0:food/00937"
          facetSids: "at.gd/2.0:fd/1200"
          amountConsumed: 11.359999656677246
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          annotations: {}
        - type: "FOOD"
          name: "Pommes frites"
          sid: "at.gd/2.0:food/00010"
          facetSids: "at.gd/2.0:fd/0204, at.gd/2.0:fd/0307, at.gd/2.0:fd/0402, at.gd/2.0:fd/0801,\
            \ at.gd/2.0:fd/1200"
          amountConsumed: 142
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/0101"
        - type: "FRYING_FAT"
          name: "Öl, Raps"
          sid: "at.gd/2.0:food/00937"
          facetSids: "at.gd/2.0:fd/1200"
          amountConsumed: 5.679999828338623
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          annotations: {}
        - type: "FOOD"
          name: "Pommes frites"
          sid: "at.gd/2.0:food/00010"
          facetSids: "at.gd/2.0:fd/0204, at.gd/2.0:fd/0309, at.gd/2.0:fd/0402, at.gd/2.0:fd/0801,\
            \ at.gd/2.0:fd/1200"
          amountConsumed: 142
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/0101"
        - type: "FRYING_FAT"
          name: "Öl, Raps"
          sid: "at.gd/2.0:food/00937"
          facetSids: "at.gd/2.0:fd/1200"
          amountConsumed: 11.359999656677246
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          annotations: {}
        - type: "FOOD"
          name: "Pommes frites"
          sid: "at.gd/2.0:food/00010"
          facetSids: "at.gd/2.0:fd/0204, at.gd/2.0:fd/0301, at.gd/2.0:fd/0402, at.gd/2.0:fd/0801,\
            \ at.gd/2.0:fd/1200"
          amountConsumed: 142
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/0101"
    - hourOfDay: "14:00:00"
      foodConsumptionOccasionId: "07"
      foodConsumptionPlaceId: "01"
      memorizedFood:
      - name: "Kuchen"
        topLevelRecords:
        - type: "FOOD"
          name: "Kuchen n.s. {assocRecp=638}"
          sid: "at.gd/2.0:food/01151"
          facetSids: "at.gd/2.0:fd/0699, at.gd/2.0:fd/0702, at.gd/2.0:fd/1001, at.gd/2.0:fd/1500,\
            \ at.gd/2.0:fd/1602"
          amountConsumed: 80
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/1201"
        - type: "FOOD"
          name: "Kuchen n.s. {assocRecp=638}"
          sid: "at.gd/2.0:food/01151"
          facetSids: "at.gd/2.0:fd/0699, at.gd/2.0:fd/0702, at.gd/2.0:fd/0801, at.gd/2.0:fd/1006"
          amountConsumed: 80
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/1201"
        - type: "FOOD"
          name: "Kuchen n.s. {assocRecp=638}"
          sid: "at.gd/2.0:food/01151"
          facetSids: "at.gd/2.0:fd/0699, at.gd/2.0:fd/0702, at.gd/2.0:fd/1009"
          amountConsumed: 80
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/1201"
      - name: "Topfentorte"
        topLevelRecords:
        - type: "FOOD"
          name: "Topfentorte {assocRecp=434}"
          sid: "at.gd/2.0:food/01179"
          facetSids: "at.gd/2.0:fd/0699, at.gd/2.0:fd/0702, at.gd/2.0:fd/1001, at.gd/2.0:fd/1500,\
            \ at.gd/2.0:fd/1602"
          amountConsumed: 88
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/1201"
        - type: "FOOD"
          name: "Topfentorte {assocRecp=434}"
          sid: "at.gd/2.0:food/01179"
          facetSids: "at.gd/2.0:fd/0699, at.gd/2.0:fd/0702, at.gd/2.0:fd/0801, at.gd/2.0:fd/1006"
          amountConsumed: 88
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/1201"
    - hourOfDay: "19:00:00"
      foodConsumptionOccasionId: "09"
      foodConsumptionPlaceId: "01"
      memorizedFood:
      - name: "Fisch, Lachs"
        topLevelRecords:
        - type: "FOOD"
          name: "Fisch, Lachs"
          sid: "at.gd/2.0:food/00872"
          facetSids: "at.gd/2.0:fd/0131, at.gd/2.0:fd/0204, at.gd/2.0:fd/0328, at.gd/2.0:fd/0499,\
            \ at.gd/2.0:fd/1001, at.gd/2.0:fd/1301, at.gd/2.0:fd/2807"
          amountConsumed: 56.549999237060547
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/0801"
        - type: "FRYING_FAT"
          name: "Öl, Raps"
          sid: "at.gd/2.0:food/00937"
          facetSids: "at.gd/2.0:fd/1200"
          amountConsumed: 1.6964999437332153
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          annotations: {}
        - type: "FOOD"
          name: "Fisch, Lachs"
          sid: "at.gd/2.0:food/00872"
          facetSids: "at.gd/2.0:fd/0131, at.gd/2.0:fd/0204, at.gd/2.0:fd/0308, at.gd/2.0:fd/0402,\
            \ at.gd/2.0:fd/0905, at.gd/2.0:fd/1009, at.gd/2.0:fd/1301, at.gd/2.0:fd/2807"
          amountConsumed: 56.549999237060547
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/0801"
        - type: "FRYING_FAT"
          name: "Öl, Raps"
          sid: "at.gd/2.0:food/00937"
          facetSids: "at.gd/2.0:fd/1200"
          amountConsumed: 2.8275001049041748
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          annotations: {}
      - name: "Gemüse"
        topLevelRecords:
        - type: "FOOD"
          name: "Gemüsemischung n.s."
          sid: "at.gd/2.0:food/00017"
          facetSids: "at.gd/2.0:fd/0204, at.gd/2.0:fd/0307, at.gd/2.0:fd/0499, at.gd/2.0:fd/0910,\
            \ at.gd/2.0:fd/1200, at.gd/2.0:fd/1302, at.gd/2.0:fd/2807"
          amountConsumed: 79
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/0200"
        - type: "FRYING_FAT"
          name: "Öl, Raps"
          sid: "at.gd/2.0:food/00937"
          facetSids: "at.gd/2.0:fd/1200"
          amountConsumed: 6.3199996948242188
          consumptionUnit: "GRAM"
          rawToCookedCoefficient: 1
          annotations: {}
    annotations:
      dataSource: "Byte-Resource[ef bb bf 3c 3f 78 6d 6c 20 76 65 72 73 69 6f 6e ...]"
annotations: {}
